### ISSUE TITLE
Support for manual cache refresh of Dashboards/Questions (non-expiring results)

### ISSUE THREAD

**Issue by gakselrod** [metabase/metabase#9983]
We have a lot of dashboards and questions that are expensive to run (1-5 minutes), typically showing long-range trends weekly or monthly metrics. These are often used to quickly pull data and context into a discussion, so it's critical to have them load **instantly** . Usually, I don't need fresh data; just something that's less than a ~month old. 

I'd like to be able to set dashboards and questions to "manual refresh", and they would cache the last result indefinitely (or for a very long time, e.g. months). If I find that it's too out of date, I'll click refresh, but usually the data there will be just fine. Redash does this well. Can we pretty pretty please add this to metabase? (I searched and couldn't find an existing issue to +1). Thank you!

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by mazameli** [comment]
Have you tried using the [caching](https://metabase.com/docs/latest/administration-guide/14-caching.html) feature?

**Comment by salsakran** [comment]
@gakselrod a better way to think about this sort of problem is to precompute just enough of the big cards' queries that you can then slice and dice the intermediate results in real-time. 

If you're ok with month old data, doing a daily or weekly tranformation would give you pretty reasonably fresh data while at the same time letting you explore + interactively query in real time.

**Comment by flamber** [comment]
Related to #6072

**Comment by gakselrod** [comment]
@mazameli Yeah we've looked at caching but we have hundreds of employees and some of them use dashboards for Ops purposes and need the dash to load live nearly every time.

@salsakran that would be ideal but we have thousands of dashes usually created by relatively novice users and they don't have access to create materialized views in the underlying databases.

**Comment by RMMSecurity** [comment]
Why not turn on caching globally and adjust the settings to be almost unreasonably high as to not hit? Then you can adjust the caching for each individual question(card) with your custom ttl. The field is under report_card named "cache_ttl". Unfortunately there's no front-end UI for it yet, but in the interim that should do it!

Source: https://github.com/metabase/metabase/blob/675840433658bf8a20ab8b6fe62f3389813dbd75/src/metabase/query_processor/middleware/cache.clj#L4-L6

**Comment by gakselrod** [comment]
I didn't know that existed! It's great that that exists. However, we really need front-end control of this for it to scale to our ops users.

Also, for queries that are being cached, we need a button to force a refresh when the user needs fresh data earlier than the TTL. 

When do we think these features might be developed?

**Comment by RMMSecurity** [comment]
@gakselrod It likely isn't on their roadmap yet. There are three ways you could possibly go about it though. You could embed Metabase and have elements on a frame to make those API calls to change the ttl values or do the refresh avoiding the cache. You could buy their Enterprise addition which involves feature requests. Or of course you could fork Metabase and add it in directly (but that's a lot of hassle)

As for what you're describing the underlying elements are all already there. Interacting with the API or database let's you set those values. The only problem with adding it into the frontend directly is their team is in the middle of a UI redesign for the GUI builder at the moment. That and grappling with how to implement it in the Metabase design style.

**Comment by theoaf** [comment]
@RMMSecurity would you be able to expand on setting the cache for individual cards?

I see that there's a `cache_ttl` column in the report_card table. I tried updating that to 100 for a query that takes 5 seconds. and it still takes 5 seconds - in other words, the caching didn't work.

am I missing something here?

**Comment by vijay-kumar-gupta** [comment]
@theoaf how to set cache_ttl for a specific question.
I have connected the redshift database.

Example:

I have a question which displays the number of orders for the current date:
Query in question:

select count* from orders where date=current_date,
now, how can I set cache_ttl on this particular question?

**Comment by theoaf** [comment]
> @theoaf how to set cache_ttl for a specific question.
> I have connected the redshift database.
> 
> Example:
> 
> I have a question which displays the number of orders for the current date:
> Query in question:
> 
> select count* from orders where date=current_date,
> now, how can I set cache_ttl on this particular question?

Hi @vijay-kumar-gupta I have a python function that looks like this:
```
def set_card_cache_ttl(card_id, ttl=3000):
    query_mbdb("""UPDATE public.report_card 
                          SET cache_ttl = """ + str(ttl) + """
                          WHERE id = """ + str(card_id))
```
`query_mbdb` simply runs the string argument on the metabase application database

**Comment by vijay-kumar-gupta** [comment]
@theoaf thanks for the response.

one more question, how can I access the Metabase application database running on AWS ec2 instance.
Can you share the commands?

**Comment by theoaf** [comment]
what are you using as your application database? H2 is the default but metabase recommend setting up a proper database for production:

https://www.metabase.com/docs/latest/operations-guide/configuring-application-database.html

We have ours running on postgres, so accessing it is like accessing any other postgres database. 

the `query_mbdb` I described uses sqlalchemy to connect and execute the sql from python.

**Comment by flamber** [comment]
@vijay-kumar-gupta Please use the forum for questions and troubleshooting: https://discourse.metabase.com/
@theoaf Thanks for helping :+1:

**Comment by vijay-kumar-gupta** [comment]
@theoaf 
I have gone through the above link.
On running ls Metabase.* as mentioned in the above link it is not showing any database files.

We have a default h2 database but I am not able to access it, any help will be much appreciated.

**Comment by luizarakaki** [comment]
It is now possible to set the cache duration in Pro/EE instances for questions and dashboards to very long times and discard the cache when needed

**Comment by alexyarosh** [comment]
reopened unintentionally
